### PR TITLE
prefix GVK labels in CustomResourceMonitoring

### PR DESCRIPTION
**Pull request by bavarianbidi**
Signed-off-by: Mario Constanti <mario@constanti.de>

**What this PR does / why we need it**:

This PR will prefix the auto-generated GVK labels for CustomResources with `cr_` to make it a bit more clear that these labels got generated.
It also make it more easier to reflect e.g. a `version` of a `CR` in a metric by just label it `version` instead of some specific `app_version` (or similar).

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by CatherineF-dev** [comment]
I think "group" is fine and it aligns with k8s documentations, cc @rexagod to have another review.

**Comment by dgrisonnet** [comment]
I am a bit on the fence here as I like the current naming because it aligns with the labels we have in the kube-apiserver HTTP metrics. But at the same time I can see how one could want to use a version label for other purposes. Maybe we could prefix them with `crd` similarly to how we prefix the metrics name `kube_crd`?

Do you have an opinion on that @logicalhan?

**Comment by bavarianbidi** [comment]
Don't get me wrong, but it's really nice to get the `group`, `version` and `kind` data as label values in the metric without additional work. 
But the main purpose from a user point of view is to reflect information from CRs. 
As many third-party-components have a `version` field which i'm more interested in than the version of the `CRD` itself (as i already have to define the GVK for a CRD in the configuration) i'm very happy if we find a solution for that.

As an alternative we could also think about an additional configuration value/flag where a user can opt-in/opt-out the auto-generated `GVK` labels.

Beside the origin issue with the `version` field,  i've already stumbled 2 times this morning about the `version` field in a `CR` where i have to decided how to continue.

from `cluster-api-azure`, where an `AzureMachinePool` can reflect the `version` of an image in the `status` field:

```
k explain azuremachinepool.status.image.marketplace.version
KIND:     AzureMachinePool
VERSION:  infrastructure.cluster.x-k8s.io/v1beta1

FIELD:    version <string>

DESCRIPTION:
     Version specifies the version of an image sku. The allowed formats are
     Major.Minor.Build or 'latest'. Major, Minor, and Build are decimal numbers.
     Specify 'latest' to use the latest version of an image available at deploy
     time. Even if you use 'latest', the VM image will not automatically update
     after deploy time even if a new version becomes available.
```

from `flux`, where you can define a `HelmRelease` and point to a specific version:

```
$ k explain helmreleases.spec.chart.spec.version
KIND:     HelmRelease
VERSION:  helm.toolkit.fluxcd.io/v2beta1

FIELD:    version <string>

DESCRIPTION:
     Version semver expression, ignored for charts from v1beta2.GitRepository
     and v1beta2.Bucket sources. Defaults to latest when omitted.

```

**Comment by dgrisonnet** [comment]
> As many third-party-components have a version field which i'm more interested in than the version of the CRD itself (as i already have to define the GVK for a CRD in the configuration) i'm very happy if we find a solution for that.

Yeah this scenario totally makes sense.

> As an alternative we could also think about an additional configuration value/flag where a user can opt-in/opt-out the auto-generated GVK labels.

I wouldn't be in favor of that since I don't think that there are any advantages to removing the GVK labels since they provide key information for barely any overhead.

File: docs/customresourcestate-metrics.md
Comment (dgrisonnet) [inline docs/customresourcestate-metrics.md:52]: I would prefer if we had some consistency between this prefix and the one in the metric name. Currently the metric name is using the `crd` prefix which might not be well suited. Might be better to go with `kube_cr` or `kube_custom_resources`.

Comment (bavarianbidi) [reply docs/customresourcestate-metrics.md:52]: do you mean we should use the future prefix in the metric name and in the label itself?
I like `kube_custom_resource` more than `kube_cr`.

**Comment by CatherineF-dev** [comment]
Another way might be kube-state-metrics to support renaming labels for CRs.
So that kube-state-metrics collects HelmRelease.version and renames it into helm_version.

```
$ k explain helmreleases.spec.chart.spec.version
KIND:     HelmRelease
VERSION:  helm.toolkit.fluxcd.io/v2beta1

FIELD:    version <string>

DESCRIPTION:
     Version semver expression, ignored for charts from v1beta2.GitRepository
     and v1beta2.Bucket sources. Defaults to latest when omitted.
```

File: docs/customresourcestate-metrics.md
Comment (dgrisonnet) [reply docs/customresourcestate-metrics.md:52]: Yes, I'd like to avoid having `crd` somewhere and `cr` elsewhere, it's better be consistent.

Comment (mrueg) [reply docs/customresourcestate-metrics.md:52]: I agree, would be nice to have one thing. I believe right now we mix Custom Resource, Custom Resource Definition and Custom Resource State.  `kube_customresource` would align with `kube_horizontalpodautoscaler` (no additional underscores).

Should we rename everything custom resource state into simply custom resource and abbreviate it with "cr" where necessary?

Comment (bavarianbidi) [reply docs/customresourcestate-metrics.md]: i've pushed a new commit (https://github.com/kubernetes/kube-state-metrics/pull/1942/commits/ed85920888a27a984f5f1a65c6e2d7caedee2948) - this should reflect the way we want to go further:

Comment (bavarianbidi) [reply docs/customresourcestate-metrics.md]: FYI: Based on the previous commit i started to improve the e2e tests because they didn't cover the default generated `group`,`version`,`kind` labels atm. will continue on monday morning.

File: pkg/customresourcestate/custom_resource_metrics_test.go
Comment (CatherineF-dev) [inline pkg/customresourcestate/custom_resource_metrics_test.go:212]: customresource_group to cr_group?

Comment (bavarianbidi) [reply pkg/customresourcestate/custom_resource_metrics_test.go:212]: comming from [this comment](https://github.com/kubernetes/kube-state-metrics/pull/1942#discussion_r1061934899)

> I agree, would be nice to have one thing. I believe right now we mix Custom Resource, Custom Resource Definition and Custom Resource State. `kube_customresource` would align with `kube_horizontalpodautoscaler` (no additional underscores).
> 
> Should we rename everything custom resource state into simply custom resource and abbreviate it with "cr" where necessary?

i've started with `customresource` (because i like it a bit more) but as i'm not very opinionated i'm also fine to switch to `cr`

File: pkg/customresourcestate/config.go
Comment (bavarianbidi) [reply pkg/customresourcestate/config.go:71]: thanks for the hint @mrueg - done

File: docs/verticalpodautoscaler-metrics.md
Comment (mrueg) [inline docs/verticalpodautoscaler-metrics.md:33]: It looks like these also use the old naming scheme which we replaced (and this MR updates its fields).

I believe they should look like this: 
```
# HELP kube_customresource_annotations Kubernetes annotations converted to Prometheus labels.
# TYPE kube_customresource_autoscaling_k8s_io_v1_VerticalPodAutoscaler_annotations gauge
# kube_customresource_annotations{customresource_group="autoscaling.k8s.io", customresource_version="v1", customresource_kind="VerticalPodAutoscaler", namespace="default",target_api_version="autoscaling.k8s.io/v1",target_kind="Deployment",target_name="hamster",verticalpodautoscaler="hamster-vpa"} 123
```

File: docs/customresourcestate-metrics.md
Comment (mrueg) [inline docs/customresourcestate-metrics.md:208]: group, kind, version should be prefixed with customresource_ in this file

File: pkg/customresourcestate/custom_resource_metrics_test.go
Comment (CatherineF-dev) [reply pkg/customresourcestate/custom_resource_metrics_test.go:212]: cc @mrueg which label name do you prefer? `cr_group` or `customresource_group`

Comment (mrueg) [reply pkg/customresourcestate/custom_resource_metrics_test.go:212]: My personal preference is the second one, two letter acronyms can become ambiguous

**Comment by bavarianbidi** [comment]
I've got 2 approvals... Who could give an LGTM to get this merged?

**Comment by mrueg** [comment]
I'll make sure to merge it by the end of this week, want to provide other maintainers the opportunity to give further feedback

**Review by dgrisonnet — APPROVED** [review]
Thank you for bearing with us @bavarianbidi :)

/lgtm
/approve

**Pull request merged**
